File: scripts/scenario.test.mjs
import assert from 'node:assert/strict';
import { cp, mkdtemp, readFile, rm, writeFile } from 'node:fs/promises';
import os from 'node:os';
import path from 'node:path';
import { spawnSync } from 'node:child_process';
import test from 'node:test';
import { fileURLToPath } from 'node:url';

const repositoryRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const definitionPath = path.join(repositoryRoot, 'testsupport', 'scenario', 'deployment-scenario.json');

test('the scenario definition has one ordered six-step smoke prefix', async () => {
  const definition = JSON.parse(await readFile(definitionPath, 'utf8'));
  assert.equal(definition.schemaVersion, 1);
  assert.equal(definition.smokeBudgetSeconds, 180);
  assert.equal(definition.steps.length, 13);
  assert.deepEqual(definition.steps.slice(0, 6).map(step => step.level), Array(6).fill('smoke'));
  assert.equal(new Set(definition.steps.map(step => step.id)).size, definition.steps.length);
  for (const step of definition.steps) {
    assert.match(step.id, /^[a-z0-9-]+$/);
    assert.ok(step.assertion.type);
    assert.notEqual(step.assertion.expected, undefined);
    assert.equal(step.evidence, `${step.id}.json`);
  }
});

test('the fixed fake CLIs make the same red repository green with the same commit', async () => {
  const root = await mkdtemp(path.join(os.tmpdir(), 'deployment-scenario-contract-'));
  const childEnvironment = { ...process.env };
  delete childEnvironment.NODE_TEST_CONTEXT;
  try {
    const shas = [];
    for (const name of ['first', 'second']) {
      const repository = path.join(root, name);
      const log = path.join(root, `${name}.log`);
      await cp(path.join(repositoryRoot, 'testsupport', 'scenario', 'fixture-repository'), repository, { recursive: true });
      let result = spawnSync('git', ['init', '-b', 'main'], { cwd: repository, encoding: 'utf8' });
      assert.equal(result.status, 0, result.stderr);
      result = spawnSync('git', ['add', '.'], { cwd: repository, encoding: 'utf8' });
      assert.equal(result.status, 0, result.stderr);
      result = spawnSync('git', ['-c', 'user.name=Deployment Scenario', '-c', 'user.email=scenario@example.invalid',
        'commit', '-m', 'test: seed deployment scenario'], {
        cwd: repository, encoding: 'utf8',
        env: { ...process.env, GIT_AUTHOR_DATE: '2026-09-06T12:00:00Z', GIT_COMMITTER_DATE: '2026-09-06T12:00:00Z' }
      });
      assert.equal(result.status, 0, result.stderr);
      assert.notEqual(spawnSync(process.execPath, ['--test', 'scenario.test.mjs'], { cwd: repository, env: childEnvironment }).status, 0);
      await writeFile(log, '', 'utf8');
      result = spawnSync(process.execPath, [path.join(repositoryRoot, 'testsupport', 'scenario', 'fake-coding-cli.mjs'), repository, log], { cwd: repository, encoding: 'utf8', env: childEnvironment });
      assert.equal(result.status, 0, result.stderr);
      result = spawnSync(process.execPath, [path.join(repositoryRoot, 'testsupport', 'scenario', 'fake-review-cli.mjs'), repository, log], { cwd: repository, encoding: 'utf8', env: childEnvironment });
      assert.equal(result.status, 0, result.stderr);
      shas.push(spawnSync('git', ['rev-parse', 'HEAD'], { cwd: repository, encoding: 'utf8' }).stdout.trim());
    }
    assert.equal(shas[0], shas[1]);
  } finally {
    await rm(root, { recursive: true, force: true });
  }
});

test('the shell runner reserves exit code 2 for invalid configuration', () => {
  const result = spawnSync('bash', [path.join(repositoryRoot, 'scripts', 'scenario.sh'), '--target', 'invalid', '--level', 'smoke'], { encoding: 'utf8' });
  assert.equal(result.status, 2);
});

test('distributed container entry points use the published assembly names', async () => {
  const expected = new Map([
    ['task-server/Dockerfile', 'ENTRYPOINT ["dotnet", "task-server.dll"]'],
    ['studio-bff/Dockerfile', 'ENTRYPOINT ["dotnet", "agent-studio-bff.dll"]'],
    ['orchestrator-engine/Dockerfile', 'ENTRYPOINT ["dotnet", "orchestrator-engine.dll"]'],
    ['testsupport/scenario/Dockerfile', 'ENTRYPOINT ["dotnet", "/opt/agent-host/agent-host.dll", "--poll"]']
  ]);
  for (const [file, line] of expected) {
    const content = await readFile(path.join(repositoryRoot, file), 'utf8');
    assert.match(content, new RegExp(line.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')));
  }
});

File: scripts/scenario_test.py
import importlib.util
import json
from pathlib import Path
import tempfile
import unittest
import xml.etree.ElementTree as ET


ROOT = Path(__file__).resolve().parents[1]
SPEC = importlib.util.spec_from_file_location("deployment_scenario", ROOT / "scripts" / "scenario.py")
scenario = importlib.util.module_from_spec(SPEC)
assert SPEC.loader is not None
SPEC.loader.exec_module(scenario)


class DeploymentScenarioContractTests(unittest.TestCase):
    def test_definition_has_six_step_smoke_prefix_and_typed_assertions(self) -> None:
        definition = json.loads(scenario.DEFINITION.read_text(encoding="utf-8"))
        scenario.validate_definition(definition)
        smoke = [step["id"] for step in definition["steps"] if "smoke" in step["levels"]]
        self.assertEqual(
            ["bootstrap-principals", "register-runner", "create-task", "claim",
             "run-fake-cli", "auto-review"],
            smoke,
        )

    def test_failure_report_is_junit_and_markdown_with_stable_failure_count(self) -> None:
        definition = {"id": "fixture", "schemaVersion": 1}
        rows = [{"id": "broken", "title": "Broken step", "status": "failed",
                 "duration": 0.125, "evidence": "evidence/broken.json", "message": "expected failure"}]
        with tempfile.TemporaryDirectory() as directory:
            output = Path(directory)
            scenario.write_reports(output, definition, "inproc", "smoke", rows, 0.0)
            suite = ET.parse(output / "scenario.junit.xml").getroot()
            self.assertEqual("1", suite.attrib["failures"])
            self.assertIn("Deployment scenario: FAIL", (output / "scenario-report.md").read_text())


if __name__ == "__main__":
    unittest.main()
